libdistopia/src/distopia.cpp

```diff
@@ -626,6 +626,64 @@ namespace distopia {
 
         }
 
+        template <typename T, typename B>
+        void CalcSelfDistanceArray(const T *a, int n, T *out, B &box) {
+            // n is *at least* nlanes
+            const hn::ScalableTag<T> d;
+            int nlanes = hn::Lanes(d);
+
+            // load all of one value into register
+            // loop over spans following that value
+
+            // for storing stub end values
+            T out_sub[HWY_MAX_LANES_D(hn::ScalableTag<T>)];
+
+            auto b_x = hn::Undefined(d);
+            auto b_y = hn::Undefined(d);
+            auto b_z = hn::Undefined(d);
+            // we are saving a triangular matrix into a flattened array
+            // this is the "row" offset in output array
+            size_t p_a = 0;
+            for (size_t ia=0; ia<n-1; ia++) {  // ignore final a as it has nothing to compare to
+                // a registers hold single value
+                auto a_x = hn::Set(d, a[3 * ia]);
+                auto a_y = hn::Set(d, a[3 * ia + 1]);
+                auto a_z = hn::Set(d, a[3 * ia + 2]);
+                // "column" offset in output array
+                size_t p_b = 0;
+                for (size_t ib = ia + 1; ib + nlanes <= n; ib += nlanes) {
+                    // b registers hold values to compare against a
+                    hn::LoadInterleaved3(d, a + 3*ib, b_x, b_y, b_z);
+
+                    auto result = box.Distance(a_x, a_y, a_z, b_x, b_y, b_z);
+                    hn::StoreU(result, d, out + p_a + p_b);
+                    p_b += nlanes;
+                }
+                // remainder loop
+                // all calls will have messy final sizes,
+                // e.g. final a iteration always has one b value to compare against
+                // therefore can't use previous backtrack strategy (calcbonds) to always handle a full vector width
+                // and can't use distancearray strategy of chickening out for small n
+                size_t iteration_size = n - 1 - ia;
+                size_t rem = (iteration_size) % nlanes;
+                if (rem) {
```

Comment:
this bit here is the alternate way of dealing with overhang.  For self_DA it's a lot cleaner to use this pattern.  If we ever want to benchmark this pattern elsewhere this is roughly what to copy